File: c/src/workbook/exercises/08_Structures/ComplexNumbers/Complex.c
/*****************************************************************************************************
 * Sample solution
 *****************************************************************************************************
 * Author: Marc Hensel, http://www.haw-hamburg.de/marc-hensel
 * Project: https://github.com/MarcOnTheMoon/coding_learners_c/
 * Copyright: 2023, Marc Hensel
 * Version: 2023.08.10
 * License: CC BY-NC-SA 4.0, see https://creativecommons.org/licenses/by-nc-sa/4.0/deed.en
 *****************************************************************************************************/

/* Include files */
#define _USE_MATH_DEFINES		// Enable mathematical constant M_PI in math.h
#include <stdio.h>
#include <math.h>

/* Structures */
struct complexNumber {
	double real;
	double imaginary;
};

/* Prototypes */
double getComplexAbsolute(struct complexNumber z);
double getComplexPhase(struct complexNumber z);
struct complexNumber addComplex(struct complexNumber a, struct complexNumber b);

/* Main function */
int main(void)
{
	struct complexNumber z1 = { 2.0, -2.0 };
	struct complexNumber z2 = { -0.5, 0.7 };
	struct complexNumber sum = addComplex(z1, z2);

	// Print complex numbers
	printf("z1 = % .1f %+.1f j\n", z1.real, z1.imaginary);
	printf("z2 = % .1f %+.1f j\n", z2.real, z2.imaginary);

	// Print absolute, phase, and sum
	printf("\nAbsolute: |z1|    = %f\n", getComplexAbsolute(z1));
	printf("Phase   : arg(z1) = %f * PI\n", getComplexPhase(z1) / M_PI);
	printf("Sum     : z1 + z2 = %.1f %+.1f j\n", sum.real, sum.imaginary);

	getchar();
	return 0;
}

/* Get absolute value of a complex number */
double getComplexAbsolute(struct complexNumber z)
{
	return sqrt(z.real * z.real + z.imaginary * z.imaginary);
}

/* Get phase in [0, 2 * pi] of a complex number */
double getComplexPhase(struct complexNumber z)
{
	double phase = ((z.real != 0.0) || (z.imaginary != 0.0)) ? atan2(z.imaginary, z.real) : 0.0;

	return (phase >= 0.0) ? phase : phase + 2.0 * M_PI;
}

/* Get the sum of two complex numbers */
struct complexNumber addComplex(struct complexNumber a, struct complexNumber b)
{
	struct complexNumber z = { a.real + b.real, a.imaginary + b.imaginary };

	return z;
}

File: java/src/workbook/src/exercises/_07_gui/e_audio/Complex.java
package exercises._07_gui.e_audio;

/** Mathematical complex number z = x + i * y.<br><br>
 *
 * Author: Marc Hensel (<a href="http://www.haw-hamburg.de/marc-hensel">link</a>)<br>
 * Project: coding_learners_java (<a href="https://github.com/MarcOnTheMoon/coding_learners_java/">link</a>)<br>
 * Copyright: 2024, Marc Hensel<br>
 * Version: 2024.01.30<br>
 * License: CC BY-NC-SA 4.0 (see <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/deed.en">link</a>)
 */
public class Complex {
	private double real, imag;

	/** Instantiate object.
	 * 
	 * @param real Real part x
	 * @param imag Imaginary part y
	 */
	public Complex(double real, double imag) {
		super();
		this.real = real;
		this.imag = imag;
	}

	/** Get the real part x.
	 * 
	 * @return real part x
	 */
	public double getReal() {
		return real;
	}

	/** Get the imaginary part y.
	 * 
	 * @return imaginary part y
	 */
	public double getImag() {
		return imag;
	}

	/** Get absolute |z|.
	 * 
	 * @return absolute value |z| = sqrt(x^2 + y^2)
	 */
	public double getAbsolute() {
		return Math.sqrt(real * real + imag * imag);
	}
	
	/** Set the real and imaginary part.
	 * 
	 * @param real Real part x
	 * @param imag Imaginary part y
	 */
	public void set(double real, double imag) {
		this.real = real;
		this.imag = imag;
	}
	
	/** Copy the contents of an array into another one.
	 * 
	 * @param copy [out] Destination array
	 * @param source [in] Source array
	 */
	public static void copy(Complex[] copy, Complex[] source) {
		if (copy.length == source.length) {
			for (int i = 0; i < copy.length; i++) {
				copy[i].real = source[i].real;
				copy[i].imag = source[i].imag;
			}
		}
	}

	/** Copy the real parts in an array of complex numbers into another array.
	 *  
	 * @param realParts [out] Destination array for the real parts
	 * @param source [in] Source array
	 */
	public static void copyReal(short[] realParts, Complex[] source) {
		if (realParts.length == source.length) {
			for (int i = 0; i < realParts.length; i++) {
				realParts[i] = (short)source[i].real;
			}
		}
	}
	
	@Override
	public Complex clone() {
		return new Complex(real, imag);
	}
}
